AX: Implement ARIA 1.3 properties aria-colindextext and aria-rowindextext on Apple platforms

dusek opened the pull request:
#### 5cb921272c01e41e5a4c8db532aaaabb9647d7c6
<pre>
AX: Implement ARIA 1.3 properties aria-colindextext and aria-rowindextext on Apple platforms
<a href="https://bugs.webkit.org/show_bug.cgi?id=257142">https://bugs.webkit.org/show_bug.cgi?id=257142</a>
<a href="https://rdar.apple.com/problem/109669421">rdar://problem/109669421</a>

Reviewed by Tyler Wilcock.

Per spec, these define human-readable alternatives to aria-colindex and aria-rowindex.
Makes WAI-ARIA idlharness WPT test fully pass now on macOS.

While at it, make DRT&apos;s handling of `null` JS strings returned by
AccessibilityUIElement::stringAttributeValue consistent with that of
WKRT, to be able to have same test expectation files / tests when
comparing and/or outputting result of `stringAttributeValue` that
returns a `null` (which is required for the new layout test in this
change).

* LayoutTests/accessibility/ARIA-reflection-expected.txt:
* LayoutTests/accessibility/ARIA-reflection.html:
* LayoutTests/accessibility/aria-table-indextext-expected.txt: Added.
* LayoutTests/accessibility/aria-table-indextext.html: Added.
* LayoutTests/accessibility/display-contents/role-row-headers-expected.txt:
* LayoutTests/accessibility/mac/aria-columnrowheaders-expected.txt:
* LayoutTests/accessibility/table-attributes-expected.txt:
* LayoutTests/accessibility/table-cell-spans-expected.txt:
* LayoutTests/accessibility/table-cells-expected.txt:
* LayoutTests/accessibility/table-multiple-tbodies-expected.txt:
* LayoutTests/accessibility/table-thead-tfoot-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/reactions/AriaMixin-string-attributes.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/dom/aria-attribute-reflection-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/wai-aria/idlharness.window-expected.txt:
* LayoutTests/platform/glib/TestExpectations: Ignored new test.
* LayoutTests/platform/ios/TestExpectations: Added new test.
* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::handleAttributeChange):
(WebCore::AXObjectCache::updateIsolatedTree):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/AccessibilityTable.cpp:
(WebCore::AccessibilityTable::isDataTable const):
* Source/WebCore/accessibility/AccessibilityTableCell.cpp:
(WebCore::AccessibilityTableCell::axColumnIndexText const):
(WebCore::AccessibilityTableCell::axRowIndexText const):
* Source/WebCore/accessibility/AccessibilityTableCell.h:
* Source/WebCore/accessibility/AccessibilityTableRow.cpp:
(WebCore::AccessibilityTableRow::axRowIndexText const):
* Source/WebCore/accessibility/AccessibilityTableRow.h:
* Source/WebCore/accessibility/AriaAttributes.idl:
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
(-[WebAccessibilityObjectWrapper accessibilityRowIndexDescription]):
(-[WebAccessibilityObjectWrapper accessibilityColumnIndexDescription]):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateNodeProperties):
(WebCore::createIsolatedObjectData):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
* Source/WebCore/accessibility/mac/CocoaAccessibilityConstants.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper ALLOW_DEPRECATED_IMPLEMENTATIONS_END]):
(-[WebAccessibilityObjectWrapper accessibilityAttributeValue:]):
* Source/WebCore/html/HTMLAttributeNames.in:
* Tools/DumpRenderTree/AccessibilityUIElement.cpp:
(stringAttributeValueCallback): FIX: handle `null` string same as WKRT (required for new test)
* Tools/DumpRenderTree/ios/AccessibilityUIElementIOS.mm:
(AccessibilityUIElement::stringAttributeValue):
* Tools/WebKitTestRunner/InjectedBundle/ios/AccessibilityUIElementIOS.mm:
(WTR::AccessibilityUIElement::stringAttributeValue):

Canonical link: <a href="https://commits.webkit.org/298210@main">https://commits.webkit.org/298210@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2e68d164fc320c28b4664e31480b6dc35e4fed35

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/114533 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/34279 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/24742 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/120697 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/65258 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/116422 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/34908 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/42839 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/87049 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/41959 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/117481 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/27816 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/102857 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/67440 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/27001 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/20983 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/64379 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/97193 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/21097 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/123904 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/41546 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/31008 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/95868 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/41923 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/99047 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95652 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24382 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/40816 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/18653 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/37592 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/41425 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/46935 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/41011 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/44318 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/42761 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on LayoutTests/accessibility/aria-table-indextext.html:
```diff
@@ -0,0 +1,88 @@
+</div>
+
+<script>
+    var testOutput = "This tests that aria-{row,col}indextext attributes in aria table/grid are working correctly.\n\n";
```
In new tests, we prefer to not indent when entering a `<script>` tag (I know a bunch of old tests don't follow this rule). This avoids a bit of rightward-drift.

dusek replied:
Thanks – I will re-do with other changes once current CI finishes (and make sure to find a recent example to follow).

dusek replied:
Test indent resolved in latest (pushed) commit.

twilco commented on Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
```diff
@@ -1376,6 +1376,30 @@ - (NSUInteger)accessibilityARIAColumnIndex
+{
+    if (![self _prepareAccessibilityCall])
+        return nil;
+    AccessibilityTableCell* tableCell = [self tableCellParent];
```
Since we use this table-cell pointer to call a "non-trivial" function (per the [Safer CPP guidelines](https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines#when-calling-a-non-trivial-member-function-on-an-object-hold-a-smart-pointer-to-the-object-on-the-stack)), I think we'll need to use a smart pointer instead. `RefPtr` would work well here!

FWIW, the non-trivial function called is `axRowIndexText()` — any virtual function is automatically non-trivial to the safer-cpp analysis.

dusek replied:
Resolved in latest (pushed) commit.

twilco reviewed (approved):
Great job!

rahimabdi commented on LayoutTests/platform/glib/TestExpectations:
```diff
@@ -888,6 +888,9 @@ accessibility/dismiss-action-returns-whether-keyevent-handled.html [ Skip ]
 accessibility/contenteditable-values.html [ Skip ]
 
+# Exposing aria-{row,col}indextext needs to be implemented on glib platforms.
+accessibility/aria-table-indextext.html [ Skip ]
```
@dusek I'm curious if this test (and its failures) should be checked in, rather than skipped? Are they causing a timeout failure?

twilco replied:
I think in this case, since the test leans so heavily on unimplemented `AccessibilityUIElement` functionality for glib (specifically in `AccessibilityUIElement::stringAttributeValue`), it's OK to skip 🙂 if the test was at least partially meaningful without said missing functionality, we could check in a failure, but I don't think that's the case here.

dusek replied:
@rahimabdi I simply followed example of other _"Implement aria- attribute"_ commits (specifically [aria-description](https://github.com/WebKit/WebKit/commit/7c414e7250fbc#diff-0408b0d0fdabfde8e9e0d7dc29b91a92455f4e0db9d16db0c7fdc7e2cdc7f8edR211) and [aria-braille*](https://github.com/WebKit/WebKit/commit/8fa0af4fbc0410155ec372865ecc4c869c82defb#diff-0408b0d0fdabfde8e9e0d7dc29b91a92455f4e0db9d16db0c7fdc7e2cdc7f8edR139)) which explicitly added new tests as `TestExpectations` `Skip`s for `glib` (or then `gtk`) and `win` (which now ignores all `accessibility` tests so not needed to add individual tests anymore) when initial support in that commit was only for Apple platforms – does that answer your question please?

dusek replied:
(And yes I would expect this test to timeout on non-Apple platforms at one point, as it does use a few `await expectAsync...` at the end.)

dusek replied:
Resolving per @twilco.

rahimabdi commented on Source/WebCore/accessibility/AccessibilityTableCell.cpp:
```diff
@@ -406,6 +406,22 @@ std::optional<unsigned> AccessibilityTableCell::axRowIndex() const
     return { };
 }
 
+String AccessibilityTableCell::axColumnIndexText() const
```
Should this return no value (i.e, { }) rather than null if missing?

dusek replied:
@rahimabdi `String { }` is the default-constructed `String` for which `.isNull()` is `true`. That's what IMHO will be returned by `attributeWithDefaultARIA` (called by `getAttribute`) if the attribute is missing (and no default value is provided by ARIA). Please let me know if that answers your concern?

rahimabdi commented on Source/WebCore/accessibility/AccessibilityTableCell.cpp:
```diff
@@ -406,6 +406,22 @@ std::optional<unsigned> AccessibilityTableCell::axRowIndex() const
+
+String AccessibilityTableCell::axRowIndexText() const
+{
+    if (String text = getAttribute(aria_rowindextextAttr); !text.isNull())
```
I think a string copy can be avoided here:

```
const AtomString& text = getAttribute(aria_rowindextextAttr);
    if (!text.isNull())
        return text;
```

twilco replied:
C++ NRVO avoids the extra copy out of the function, so both your suggestion and the code as-is will only have one 🙂

dusek replied:
Marking as resolved per @twilco (no change).

twilco commented on Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
```diff
@@ -806,6 +806,8 @@ - (NSArray *)accessibilityAttributeNames
         [tempArray addObject:NSAccessibilityRowHeaderUIElementsAttribute];
         [tempArray addObject:NSAccessibilityARIAColumnIndexAttribute];
         [tempArray addObject:NSAccessibilityARIARowIndexAttribute];
+        [tempArray addObject:NSAccessibilityColumnIndexDescriptionAttribute];
+        [tempArray addObject:NSAccessibilityRowIndexDescriptionAttribute];
```
Something I just realized — shouldn't NSAccessibilityRowIndexDescriptionAttribute be a supported attribute for table rows too, as `aria-rowindextext` is valid for row roles?

dusek replied:
`aria-rowindex` is valid for rows too per [spec](https://w3c.github.io/aria/#aria-rowindex), however we are also not exposing that one on rows ATM on macOS (nor iOS). I followed example of `NSAccessibilityARIARowIndexAttribute` (i.e. `aria-rowindex`) which is exposed only on cells. Please let me know what you think based on this.

twilco replied:
Ahh, good point! Seems like a great issue to fix in a follow-up PR — making rows claim support for both NSAccessibilityARIARowIndexAttribute and NSAccessibilityRowIndexDescriptionAttribute 👍 maybe we can add a FIXME next to `tableRowAttrs` noting the fact that both of these attributes are missing.

dusek replied:
Resolved in latest (pushed) commit – added FIXME mentioning a bug I filed to track it.

dusek committed:
AX: Implement ARIA 1.3 properties aria-colindextext and aria-rowindextext on Apple platforms
https://bugs.webkit.org/show_bug.cgi?id=257142
rdar://problem/109669421

Reviewed by Tyler Wilcock.

Per spec, these define human-readable alternatives to aria-colindex and aria-rowindex.
Makes WAI-ARIA idlharness WPT test fully pass now on macOS.

While at it, make DRT's handling of `null` JS strings returned by
AccessibilityUIElement::stringAttributeValue consistent with that of
WKRT, to be able to have same test expectation files / tests when
comparing and/or outputting result of `stringAttributeValue` that
returns a `null` (which is required for the new layout test in this
change).

* LayoutTests/accessibility/ARIA-reflection-expected.txt:
* LayoutTests/accessibility/ARIA-reflection.html:
* LayoutTests/accessibility/aria-table-indextext-expected.txt: Added.
* LayoutTests/accessibility/aria-table-indextext.html: Added.
* LayoutTests/accessibility/display-contents/role-row-headers-expected.txt:
* LayoutTests/accessibility/mac/aria-columnrowheaders-expected.txt:
* LayoutTests/accessibility/table-attributes-expected.txt:
* LayoutTests/accessibility/table-cell-spans-expected.txt:
* LayoutTests/accessibility/table-cells-expected.txt:
* LayoutTests/accessibility/table-multiple-tbodies-expected.txt:
* LayoutTests/accessibility/table-thead-tfoot-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/reactions/AriaMixin-string-attributes.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/dom/aria-attribute-reflection-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/wai-aria/idlharness.window-expected.txt:
* LayoutTests/platform/glib/TestExpectations: Ignored new test.
* LayoutTests/platform/ios/TestExpectations: Added new test.
* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator<<):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::handleAttributeChange):
(WebCore::AXObjectCache::updateIsolatedTree):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/AccessibilityTable.cpp:
(WebCore::AccessibilityTable::isDataTable const):
* Source/WebCore/accessibility/AccessibilityTableCell.cpp:
(WebCore::AccessibilityTableCell::axColumnIndexText const):
(WebCore::AccessibilityTableCell::axRowIndexText const):
* Source/WebCore/accessibility/AccessibilityTableCell.h:
* Source/WebCore/accessibility/AccessibilityTableRow.cpp:
(WebCore::AccessibilityTableRow::axRowIndexText const):
* Source/WebCore/accessibility/AccessibilityTableRow.h:
* Source/WebCore/accessibility/AriaAttributes.idl:
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
(-[WebAccessibilityObjectWrapper accessibilityRowIndexDescription]):
(-[WebAccessibilityObjectWrapper accessibilityColumnIndexDescription]):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateNodeProperties):
(WebCore::createIsolatedObjectData):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
* Source/WebCore/accessibility/mac/CocoaAccessibilityConstants.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper ALLOW_DEPRECATED_IMPLEMENTATIONS_END]):
(-[WebAccessibilityObjectWrapper accessibilityAttributeValue:]):
* Source/WebCore/html/HTMLAttributeNames.in:
* Tools/DumpRenderTree/AccessibilityUIElement.cpp:
(stringAttributeValueCallback): FIX: handle `null` string same as WKRT (required for new test)
* Tools/DumpRenderTree/ios/AccessibilityUIElementIOS.mm:
(AccessibilityUIElement::stringAttributeValue):
* Tools/WebKitTestRunner/InjectedBundle/ios/AccessibilityUIElementIOS.mm:
(WTR::AccessibilityUIElement::stringAttributeValue):

Canonical link: https://commits.webkit.org/298210@main
